File: ruby/testcode/codeinj_040.rb
require_relative 'shared'

TRANSFORMS = {
  'upcase'  => ->(s) { s.upcase },
  'reverse' => ->(s) { s.reverse },
  'strip'   => ->(s) { s.strip },
  'chomp'   => ->(s) { s.chomp }
}.freeze

# vuln-code-snippet start ruby_codeinj_hardcoded_transform
def apply_transform(req)
  fn = req.param('fn')
  input = req.param('input')
  transform = TRANSFORMS.fetch(fn) { return BenchmarkResponse.bad_request('unknown transform') }
  result = transform.call(input) # vuln-code-snippet safe-line ruby_codeinj_hardcoded_transform
  BenchmarkResponse.json({ result: result })
end
# vuln-code-snippet end ruby_codeinj_hardcoded_transform

File: php/testcode/codeinj_040.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_codeinj_int_keyed_handlers
function codeinj040(BenchmarkRequest $req): BenchmarkResponse {
    $handlers = [0 => 'trim', 1 => 'strtolower'];
    $id = intval($req->param('id'));
    $fn = $handlers[$id] ?? 'trim';
    $result = $fn($req->param('value')); // vuln-code-snippet safe-line php_codeinj_int_keyed_handlers
    return BenchmarkResponse::ok((string) $result);
}
// vuln-code-snippet end php_codeinj_int_keyed_handlers
